### PR TITLE
Use direct optimization for fitting refinement

### PR DESCRIPTION
**Pull request by aarmey**

### CODE REVIEW

File: syserol/tensor.py [hunk 1]
```diff
@@ -181,18 +185,73 @@ def perform_CMTF(tOrig=None, mOrig=None, r=8):
...
+
+def cost(pIn, tOrig, mOrig, r):
+    tFac = buildTensors(pIn, tOrig, mOrig, r)
+    tDiff = np.nan_to_num(tOrig - tl.kruskal_to_tensor(tFac))
+    mDiff = np.nan_to_num(mOrig - buildGlycan(tFac))
+
+    # Overall cost
+    cost = 0.5 * (np.linalg.norm(tDiff) + np.linalg.norm(mDiff))
+
+    gtFac = deepcopy(tFac)
```
Comment (aarmey) [inline syserol/tensor.py]: This sets up the gradient.

File: syserol/tensor.py [hunk 2]
```diff
@@ -181,18 +185,73 @@ def perform_CMTF(tOrig=None, mOrig=None, r=8):
...
+
+    return cost, cp_to_vec(gtFac)
+
+
+def fit_refine(tFac, tOrig, mOrig):
+    """ Refine the factorization with direct optimization. """
+    r = tFac.rank
+
+    x0 = cp_to_vec(tFac)
+    res = minimize(cost, x0, jac=True, args=(tOrig, mOrig, r), options={"disp": True, "maxiter": 3000})
```
Comment (aarmey) [inline syserol/tensor.py]: The reason I suspect the gradients are wrong is this exits after 0 iterations. I also compared my gradient to numerical differencing with [`scipy.optimize.approx_fprime`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.approx_fprime.html).

**Comment by aarmey** [comment]
Might have everything working here.

**Comment by cyrillustan** [comment]
Great! I'm still reading the paper/code though.

**Pull request merged**